msdbook/tests/test_utils.py

```diff
@@ -0,0 +1,214 @@
+
+    # Dynamically generate grid and levels based on input data to reflect the data range
+    xgrid_min, xgrid_max = sample_data['Predictor1'].min(), sample_data['Predictor1'].max()
+    ygrid_min, ygrid_max = sample_data['Predictor2'].min(), sample_data['Predictor2'].max()
+    xgrid = np.linspace(xgrid_min - 1, xgrid_max + 1, 50)
+    ygrid = np.linspace(ygrid_min - 1, ygrid_max + 1, 50)
+    levels = np.linspace(0, 1, 10)
+    
+    contour_cmap = 'viridis'
+    dot_cmap = 'coolwarm'
+    
+    # Call the plot function
+    contourset = plot_contour_map(
+        ax,
+        result,
+        sample_data,
+        contour_cmap,
+        dot_cmap,
+        levels,
+        xgrid,
+        ygrid,
+        'Predictor1',
+        'Predictor2',
+        base=0,
+    )
+
+    # Verify the plot and axis limits/labels are correct
+    assert contourset is not None
+    assert ax.get_xlim() == (xgrid.min(), xgrid.max())
+    assert ax.get_ylim() == (ygrid.min(), ygrid.max())
+    assert ax.get_xlabel() == 'Predictor1'
+    assert ax.get_ylabel() == 'Predictor2'
+
+    # Verify that scatter plot is present by checking the number of points
+    assert len(ax.collections) > 0  
+    plt.close(fig)
+
+
+def test_empty_data():
+    """Test with empty data to ensure no errors."""
+    empty_df = pd.DataFrame({
+        'Success': [],
+        'Predictor1': [],
+        'Predictor2': [],
+        'Interaction': []
+    })
+    
+    # Test if fitting with empty data raises an error
+    with pytest.raises(ValueError):
+        fit_logit(empty_df, ['Predictor1', 'Predictor2'])
+
+    # Test plotting with empty data (skip fitting since it's empty)
+    fig, ax = plt.subplots()
+
+    # Ensure that no fitting occurs on an empty DataFrame
+    if not empty_df.empty:
+        result = fit_logit(empty_df, ['Predictor1', 'Predictor2'])
+        contourset = plot_contour_map(
+            ax, result, empty_df,
+            'viridis', 'coolwarm', np.linspace(0, 1, 10), np.linspace(-2, 2, 50),
+            np.linspace(-2, 2, 50), 'Predictor1', 'Predictor2', base=0
+        )
+        assert contourset is not None
+    else:
+        # Skip plotting if DataFrame is empty
+        assert True  # Ensures that we expect no result or plotting for empty DataFrame
+    
+    plt.close(fig)
+
+
+
+def test_invalid_predictors(sample_data):
+    """Test with invalid predictors."""
+    invalid_predictors = ['InvalidPredictor1', 'InvalidPredictor2']
+    
+    with pytest.raises(KeyError):
+        fit_logit(sample_data, invalid_predictors)
+
+
+def test_logit_with_interaction(sample_data):
+    """Test logistic regression with interaction term."""
+    sample_data['Interaction'] = sample_data['Predictor1'] * sample_data['Predictor2']
+    result = fit_logit(sample_data, ['Predictor1', 'Predictor2'])
+    
+    # Ensure the interaction term is included in the result
+    assert 'Interaction' in result.params.index
+
+
+def test_fit_logit_comprehensive(sample_data):
```

Comment:
Can you add this to the parameterized tests (checking for `result.params.values`, `result.pvalues`)